identity-iota/src/credential/credential_validator.rs

```diff
@@ -0,0 +1,645 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::credential::errors::AccumulatedCredentialValidationError;
+use crate::did::IotaDID;
+use crate::document::ResolvedIotaDocument;
+use crate::Error;
+use crate::Result;
+use identity_core::common::OneOrMany;
+use identity_core::common::Timestamp;
+use identity_credential::credential::Credential;
+use identity_did::verifiable::VerifierOptions;
+use serde::Serialize;
+
+use super::errors::ValidationError;
+use super::CredentialValidationOptions;
+
+/// A struct for validating [Credential]s.
+#[non_exhaustive]
+pub struct CredentialValidator;
+
+// Use the following pattern throughout: for each public method there is corresponding private method returning a more
+// local error.
```

Comment:
`full_validation` (or more precisely `full_validation_local_error`) calls all of these validation units and all encountered errors are collected into a `ValidationError` container  (two containers in the case of `PresentationValidator::full_validation `Vec` and `BTreeMap`).  Without the local_error variants of these validation units there would have to be a lot more (most likely bug prone) logic in the `full_validation` methods. I would be happy to revisit this code when the [never type](https://rust-lang.github.io/never-type-initiative/index.html) gets stabilized as I am hoping things like this might become possible then:
``` 
let validation_error: ValidationError = match Err(error) {
    Error::UnsuccessfulValidationUnit(err) => err,
    _ => ! 
 };
 ```
i.e. the compiler will be able to infer that certain variants cannot occur.  Since the *_local_error functions are private it is easy to remove them in the future if they are no longer needed, but right now I think they do more good than harm.

Reply:
If the only difference is in what type of error is returned from the inner versus the outer function then I rate we just return the inner `ValidationUnitResult` rather than converting it into an `identity_iota::Error`?

The smaller error enum is also more specific for developers to see the possible errors too. Not sure how that fits into the error handling guidelines but it seems good to me and it would reduce the boilerplate.

Reply:
Also not sure how just returning the inner `ValidationUnitResult` fits with the new error guidelines. However looking at it in isolation I would say that I expect most callers to prefer `ValidationUnitResult` (and `std::Result<(),AccumulatedCredentialValidationError>`, `std::Result<(),CompoundPresentationValidationError`>) over `Result<()>`.  This is because I expect callers are mostly interested in matching on these errors over simply logging them: If the issuer was not trusted they want to ask the end user for another credential from another issuer, if the issuance date is wrong the end user will want to know when they can try again etc. Matching on these variants becomes easier when one does not have to go through the crate level `Error` first.  The additional boilerplate of going through the crate level error is demonstrated in the tests for `CredentialValidator` and `PresentationValidator`. 

If we are to go with making these public methods return the local errors, we should probably also make sure that they implement  `Into<identity_iota::Error>` so callers can easily wrap them into the same error as everything else from this crate if they wish to do so. The [`time` crate does something like that](https://docs.rs/time/0.3.7/time/error/index.html).  

Any thoughts on this? Is it OK to somewhat stray from the now established [error guidelines](https://github.com/iotaledger/identity.rs/issues/534) in this instance and does that mean we might allow for it elsewhere where there are good arguments to do so?

Reply:
I don't think there's anything against doing this in the error guidelines, and it's a pattern we use elsewhere for trait objects like `Diff` (currently) and `Storage` (potentially).

> implement `Into<identity_iota::Error>`

Yup, that's still needed in my opinion. Does using `#[from] ValidationUnitError` with `thiserror` not handle it already though?

Reply:
> Does using #[from] ValidationUnitError with thiserror not handle it already though? 

That one is already handled. I was mostly thinking about `AccumulatedCredentialValidationError` and `CompoundPresentationError`.

Reply:
OK. I have decided to remove the `*_local_error` methods and let the public methods just return the local errors directly in this case.